### ISSUE TITLE
[dynamo] Better determinism of `ConfigModule` by walking using pytree

### ISSUE THREAD

**Issue by jon-chuang** [pytorch/pytorch#111754]
### 🚀 The feature, motivation and pitch

https://github.com/pytorch/pytorch/pull/111318

Currently, validation only occurs at the root. However, we should walk the pytree of each object to ensure types are respected.

In particular, we can do conversions of unfriendly types
- function objects into `"f{__module__}{__name__}"` strings
- sort sets and walk their objects to make them deterministic. 
    - There isn't a canonical order, however; sets have `__hash__`, but this hash is based on `id` which is non-deterministic. So sorting by hash is non-deterministic. Further, not all objects implement `>` operator. 
    - In other words, there is no feasible way to make deterministic representations of sets.

**Comment by zhxchen17** [comment]
Hi @jon-chuang I'm cleaning up the old issues and I saw you're assigned for this feature proposal. I wonder if you plan to work on this proposal recently? Thanks for bringing up the idea.
